ldelf/ta_elf_rel.c

```diff
@@ -433,8 +433,17 @@ static void e64_process_tls_tprel_rela(const Elf64_Sym *sym_tab,
 	*where = symval + mod->tls_tcb_offs + rela->r_addend;
 }
 
-/* Helper function written in assembly due to the calling convention */
-long tlsdesc_resolve(long);
+/*
+ * Helper function written in assembly due to the calling convention: all
+ * registers except x0 and x1 have to be preserved.
+ * See https://www.fsfla.org/~lxoliva/writeups/TLS/RFC-TLSDESC-ARM.txt section
+ * "Resolvers' Calling Convention". The document applies to 32-bit Arm but other
+ * sources mention similar constraints for other architectures.
+ */
+static void tlsdesc_resolve(void)
```

Comment:
I think this should be `__naked` and explicitly code the return instruction in order to guarantee that the compiler doesn't sneak in some other instruction.

Is it possible to define a the proper argument and return value? I don't suppose it will affect anything, but it's nice to see what the caller expects.

Reply:
> I think this should be `__naked` and explicitly code the return instruction in order to guarantee that the compiler doesn't sneak in some other instruction.

I thought the same, but:
```
ldelf/ta_elf_rel.c:444:1: warning: ‘naked’ attribute directive ignored [-Wattributes]
```
It looks like this directive is architecure-dependant and Aarch64 GCC doesn't support it :-/
Clang accepts it just fine however.

> Is it possible to define a the proper argument and return value? I don't suppose it will affect anything, but it's nice to see what the caller expects.

This would work with both GCC and Clang with no compilation warning:

```C
static long tlsdesc_resolve(struct tlsdesc *td)
{
	register long x0 asm ("x0") = (long)td;

	asm volatile ("ldr x0, [x0, #8]");
	return x0;
}
```

Another option is this (2):
```C
/* In compiler.h */
#ifdef __clang__
#define __naked __attribute__((naked))
#else
#define __naked
#endif

/*
 * static long tlsdesc_resolve(struct tlsdesc *td);
 */
static void __naked tlsdesc_resolve(void)
{
	asm volatile ("ldr x0, [x0, #8]");
}
```

Note that with (1) we can't add a `naked` to the Clang version ("non-ASM statement in naked function is not supported").

What do you prefer, (1) or (2)? Or something else? ;-)

Reply:
Hmm, it seems that the original version was the best. Sorry for the noise.

Reply:
OK, no worries. I have moved the `struct tlsdesc` definition in the C code instead of having it in a comment, looks better IMO. And the references are added to the assembly file.